rustworkx-core/src/coloring.rs

```diff
@@ -0,0 +1,86 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use std::cmp::Reverse;
+use std::hash::Hash;
+
+use crate::dictmap::*;
+use hashbrown::{HashMap, HashSet};
+use petgraph::visit::{EdgeRef, IntoEdges, IntoNodeIdentifiers, NodeCount};
+use rayon::prelude::*;
+
+/// Color a graph using a greedy graph coloring algorithm.
+///
+/// This function uses a `largest-first` strategy as described in [1]_ and colors
+/// the nodes with higher degree first.
+///
+/// Arguments:
+///
+/// * `graph` - The graph object to run the algorithm on
+///
+/// # Example
+/// ```rust
+///
+/// use petgraph::graph::Graph;
+/// use petgraph::graph::NodeIndex;
+/// use petgraph::Undirected;
+/// use rustworkx_core::dictmap::*;
+/// use rustworkx_core::coloring::greedy_color;
+///
+/// let g = Graph::<(), (), Undirected>::from_edges(&[(0, 1), (0, 2)]);
+/// let colors = greedy_color(&g);
+/// let mut expected_colors = DictMap::new();
+/// expected_colors.insert(NodeIndex::new(0), 0);
+/// expected_colors.insert(NodeIndex::new(1), 1);
+/// expected_colors.insert(NodeIndex::new(2), 1);
+/// assert_eq!(colors, expected_colors);
+/// ```
+///
+///
+/// .. [1] Adrian Kosowski, and Krzysztof Manuszewski, Classical Coloring of Graphs,
+///     Graph Colorings, 2-19, 2004. ISBN 0-8218-3458-4.
+pub fn greedy_color<G>(graph: G) -> DictMap<G::NodeId, usize>
```

Comment:
I have changed the name of the internal rust function to `greedy_node_color`. I don't feel strongly one way or another, but it might be slightly more clear if we are planning to add `greedy_edge_color` later.